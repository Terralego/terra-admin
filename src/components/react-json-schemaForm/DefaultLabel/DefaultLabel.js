import React from 'react';
import PropTypes from 'prop-types';
import DefaultDescriptionField from '@rjsf/core/lib/components/fields/DescriptionField';
import {
  isMultiSelect,
  getUiOptions,
  isFilesArray,
} from '@rjsf/core/lib/utils';

const REQUIRED_FIELD_SYMBOL = '*';

const Label = ({ label, required, id }) => !!label && (
  <label className="control-label" htmlFor={id}>
    {label}
    {required && <span className="required">{REQUIRED_FIELD_SYMBOL}</span>}
  </label>
);

const displayLabel = ({
  schema,
  uiSchema,
  definitions,
}) => {
  const { forceLabelDisplay } = getUiOptions(uiSchema);
  if (forceLabelDisplay) {
    return true;
  }
  if (schema.type === 'array') {
    return isMultiSelect(schema, definitions) || isFilesArray(schema, uiSchema, definitions);
  }
  if (schema.type === 'object') {
    return false;
  }
  if (schema.type === 'boolean' && !uiSchema['ui:widget']) {
    return false;
  }
  return true;
};

const DefaultLabel = props => {
  if (!displayLabel(props)) {
    return null;
  }

  const {
    schema,
    id,
    required,
    uiSchema,
    name,
    fields = {},
    formContext,
  } = props;

  const label = uiSchema['ui:title'] || schema.title || name;
  const description = uiSchema['ui:description'] || schema.description;

  const { DescriptionField = DefaultDescriptionField } = fields;

  return (
    <>
      <Label label={label} id={id} required={required} />
      {description && (
        <DescriptionField
          id={`${id}__description`}
          description={description}
          formContext={formContext}
        />
      )}
    </>
  );
};

export default DefaultLabel;


displayLabel.PropTypes = {
  schema: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    type: PropTypes.oneOf(['string, boolean', 'array', 'object', undefined]),
  }).isRequired,
  name: PropTypes.string.required,
  uiSchema: PropTypes.shape({
    'ui:title': PropTypes.string,
    'ui:description': PropTypes.string,
  }),
  fields: PropTypes.shape({}),
  formContext: PropTypes.shape({}),
};

displayLabel.defaultProps = {
  uiSchema: {
    'ui:title': '',
    'ui:description': '',
  },
  fields: {},
  formContext: {},
};
