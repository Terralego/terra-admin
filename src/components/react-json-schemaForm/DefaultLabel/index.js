import DefaultLabel from './DefaultLabel';

export default DefaultLabel;
