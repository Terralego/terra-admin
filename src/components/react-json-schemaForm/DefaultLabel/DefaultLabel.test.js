import React from 'react';
import renderer from 'react-test-renderer';
import DefaultLabel from './DefaultLabel';


jest.mock('@rjsf/core/lib/components/fields/DescriptionField', () => () => <div>Description</div>);

jest.mock('@rjsf/core/lib/utils', () => ({
  isMultiSelect: (schema, bool = true) => bool,
  getUiOptions: () => true,
  isFilesArray: (schema, uiSchema, bool = true) => bool,
}));

const props = {
  schema: {
    title: 'Foo Label',
    description: 'my foo label',
    type: 'string',
  },
  name: 'foo',
  uiSchema: {
    'ui:title': '',
    'ui:description': '',
  },
  fieds: {},
};

it('should render correctly', () => {
  const tree = renderer.create((
    <DefaultLabel
      {...props}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('should not render for multiple reasons', () => {
  it('should not render if the schema type is an object', () => {
    const tree = renderer.create((
      <DefaultLabel
        {...props}
        schema={{
          ...props.schema,
          type: 'object',
        }}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not render if the schema type is a boolean', () => {
    const tree = renderer.create((
      <DefaultLabel
        {...props}
        schema={{
          ...props.schema,
          type: 'boolean',
        }}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not render if the schema type is an array and not multiselect', () => {
    const tree = renderer.create((
      <DefaultLabel
        {...props}
        definition
        schema={{
          ...props.schema,
          type: 'array',
        }}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
