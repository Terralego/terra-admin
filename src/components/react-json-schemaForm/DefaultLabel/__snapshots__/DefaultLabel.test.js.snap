// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render for multiple reasons should not render if the schema type is a boolean 1`] = `null`;

exports[`should not render for multiple reasons should not render if the schema type is an array and not multiselect 1`] = `
Array [
  <label
    className="control-label"
  >
    Foo Label
  </label>,
  <div>
    Description
  </div>,
]
`;

exports[`should not render for multiple reasons should not render if the schema type is an object 1`] = `null`;

exports[`should render correctly 1`] = `
Array [
  <label
    className="control-label"
  >
    Foo Label
  </label>,
  <div>
    Description
  </div>,
]
`;
